### PR TITLE
`ultralytics 8.0.188` fix .grad attribute leaf Tensor Warning

### PR DESCRIPTION
**Pull request by glenn-jocher**
@Laughing-q this should fix the warning. I've re-added the smart_inference_mode() manager to model.predict() but also updated it to act as a pass through if inference_mode is already running (to avoid nesting them).

### Summary
🚀🧠🛠️

The pull request introduces a `smart_inference_mode` decorator to the `Model` class in `ultralytics/engine/model.py` to optimize inference performance and memory usage. It also improves the `decorate` function in `ultralytics/utils/torch_utils.py` to avoid unnecessary context managers and enhance code readability.

> _To make the model run faster and leaner_
> _We added a `smart_inference_mode` cleaner_
> _It checks the context_
> _And avoids any conflict_
> _With the torch version or the `no_grad` screener_

### Walkthrough
*  Decorate the `Model.forward` method with `smart_inference_mode` to improve performance and memory efficiency in inference mode ([link](https://github.com/ultralytics/ultralytics/pull/5094/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00R14), [link](https://github.com/ultralytics/ultralytics/pull/5094/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00R202))
*  Modify the `decorate` function inside `smart_inference_mode` to avoid unnecessary nesting of torch context managers and handle existing inference contexts ([link](https://github.com/ultralytics/ultralytics/pull/5094/files?diff=unified&w=0#diff-9bd07d929cb8522458c879bd81a82d9af468ffbc215e762f39b2a50a601758d7L47-R52))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements in model testing, documentation, and utility functions for better usability.

### 📊 Key Changes
- Updated the banner link in the documentation to a new URL.
- Added a code block to `tutorial.ipynb` for installing the package from the source.
- Added multiple device prediction testing in `tests/test_cuda.py`.
- Increased version to `8.0.188` in `__init__.py`.
- Reorganized utility imports in `model.py`.
- Adjusted prediction and validation methods within the `model.py` and `validator.py`.
- Improved image processing in FastSAM prompts (`prompt.py`), focusing on color and plot saving.
- Introduced annotations for disabling gradients to improve inference in `autobackend.py`.
- Updated YAML utility to allow for a broader range of data types.
- Adjusted requirements parsing regex for better package handling in `checks.py`.
- Implemented a smart check for `torch.inference_mode` to prevent redundancy in `torch_utils.py`.

### 🎯 Purpose & Impact
- Improved the user experience by streamlining installation instructions and updating links to relevant resources.
- Simplified the codebase for easier maintenance and improved testing capabilities 🛠️, which could lead to more stable builds.
- Enhanced image handling and visualization, leading to better debug and display of results 🖼️.
- Increased efficiency in model operations by managing gradients more effectively ⚡.
- Broadened the compatibility of the YAML utility function, allowing more flexible configurations 📝.
- Made requirement parsing more robust to prevent installation issues 📦.
- Optimized inference checks to align with the latest `torch` versions, potentially boosting performance 🚀.

### CODE REVIEW

**Comment by Laughing-q** [comment]
@glenn-jocher 
```python
from ultralytics import YOLO
model = YOLO('yolov8n.pt')
model = model.cpu()  # new method along with .cuda(), .half(), .float(), etc.
results = model()  # cpu inference

model = model.to('cuda:0')
results2 = model()  # cuda inference
print(model.device)  # prints cuda
```
encountered this error after switching device with this branch. It works fine with main branch.
pic-230926-1709-09

**Comment by glenn-jocher** [comment]
@Laughing-q thanks for the debug script! I think I fixed the issue, but I can't even run your script with `main` branch in Colab, I get a strange error even though GPU 0 is available initially. After running the code only CPU inference is available. I think there is a device handling bug as you mentioned before (separate from this leaf warning that should be fixed now).

Screenshot 2023-09-26 at 11 45 10

**Comment by glenn-jocher** [comment]
If I uncomment the CPU predict then CUDA predict works correctly:
Screenshot 2023-09-26 at 11 48 34

**Comment by Laughing-q** [comment]
@glenn-jocher yeah it's like there's something inside pytorch happened when the model switches device.

**Comment by Laughing-q** [comment]
> @Laughing-q thanks for the debug script! I think I fixed the issue, but I can't even run your script with `main` branch in Colab, I get a strange error even though GPU 0 is available initially. After running the code only CPU inference is available. I think there is a device handling bug as you mentioned before (separate from this leaf warning that should be fixed now).
> 
> Screenshot 2023-09-26 at 11 45 10

oh that, I manually fixed this by replacing `str(self.device)` to `self.device` before I tested the script.
https://github.com/ultralytics/ultralytics/blob/7f78fad8baceb41ba5e3f5af7b85b3e83930dd12/ultralytics/engine/model.py#L369
to:
```python
        self.overrides['device'] = self.device  # i.e. device(type='cuda', index=0) -> 'cuda:0'
```

**Comment by glenn-jocher** [comment]
@Laughing-q I get the same error now in a new GPU test I added with your code:
https://github.com/ultralytics/ultralytics/actions/runs/6312641972/job/17139101527?pr=5094

```python
@pytest.mark.skipif(not CUDA_IS_AVAILABLE, reason='CUDA is not available')
def test_predict_multiple_devices():
    model = YOLO('yolov8n.pt')
    model = model.cpu()
    _ = model()  # CPU inference

    model = model.to('cuda:0')
    _ = model()  # CUDA inference
    assert str(model.device) == 'cuda:0'

    model = model.cpu()
    _ = model()  # CPU inference
    assert str(model.device) == 'cpu'
```

**Comment by glenn-jocher** [comment]
@Laughing-q it works now! I've removed the warning and the test runs correctly. But I think passing device as an actual torch.device might break other methods like training, I'll check.

**Comment by glenn-jocher** [comment]
@Laughing-q yeah but with using torch.device in args though:

```python
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
model = model.cuda()
model.train(data='coco8.yaml', epochs=3)
```

**Comment by Laughing-q** [comment]
@glenn-jocher awesome! I just tested training and it works smoothly!

**Comment by glenn-jocher** [comment]
@Laughing-q PR merged! I expanded the tests also to test both `.cuda()` and `.to('cuda:0')`
